### PR TITLE
AK: Implement IPV4 host URL parsing to specification

### PR DESCRIPTION
**Pull request by shannonbooth**

### CODE REVIEW

File: AK/URLParser.cpp [hunk 1]
```diff
@@ -44,10 +46,184 @@ static Optional<DeprecatedString> parse_opaque_host(StringView input)
     return URL::percent_encode(input, URL::PercentEncodeSet::C0Control);
 }
 
-static Optional<DeprecatedString> parse_ipv4_address(StringView input)
+struct ParsedIPV4Number {
```
Comment (trflynn89) [inline AK/URLParser.cpp:49]: A bit nit-picky but let's be consistent in the capitalization of the `V` - the spec uses `IPv4` so let's match that in struct names like this.

File: AK/URLParser.cpp [hunk 2]
```diff
@@ -44,10 +46,184 @@ static Optional<DeprecatedString> parse_opaque_host(StringView input)
...
+    // 7. If any but the last item in numbers is greater than 255, then return failure.
+    for (size_t i = 0; i < numbers.size(); ++i) {
+        if (numbers[i] > 255) {
+            report_validation_error();
+            if (i != numbers.size() - 1)
+                return {};
+        }
+    }
+
+    // 8. If the last item in numbers is greater than or equal to 256(5 − numbers’s size), then return failure.
```
Comment (trflynn89) [inline AK/URLParser.cpp:177]: Copying superscript spec text is a bit awkward, but we tend to insert a caret `^` to make it clearer:
```
256^(5 − numbers’s size)
```

File: AK/URLParser.cpp [hunk 3]
```diff
@@ -44,10 +46,184 @@ static Optional<DeprecatedString> parse_opaque_host(StringView input)
...
+
+    // 10. Remove the last item from numbers.
+    numbers.take_last();
+
+    // 11. Let counter be 0.
+    u8 counter = 0;
+
+    // 12. For each n of numbers:
+    for (u32 n : numbers) {
+        // 1. Increment ipv4 by n × 256(3 − counter).
```
Comment (trflynn89) [inline AK/URLParser.cpp:192]: Same here (caret)

File: AK/URLParser.cpp [hunk 4]
```diff
@@ -44,10 +46,184 @@ static Optional<DeprecatedString> parse_opaque_host(StringView input)
...
+        ipv4 += n * pow<size_t>(256, 3 - counter);
```
Comment (trflynn89) [inline AK/URLParser.cpp:193]: Probably isn't going to show up in profiles, but the `pow<size_t>` invocations in this file *could* be replaced with bitwise math
```c++
ipv4 |= n << (8 * (3 - counter))
```
(not asking you to replace this here, just making a note if this does become hot)

File: AK/URLParser.cpp [hunk 5]
```diff
@@ -382,8 +564,22 @@ static Optional<DeprecatedString> parse_host(StringView input, bool is_not_speci
         }
     }
 
-    auto ipv4_host = parse_ipv4_address(ascii_domain);
-    return ipv4_host;
+    // 8. If asciiDomain ends in a number, then return the result of IPv4 parsing asciiDomain.
+    if (is_ascii_digit(ascii_domain[ascii_domain.length() - 1])) {
+
```
Comment (trflynn89) [inline AK/URLParser.cpp:569]: Don't need this newline

File: AK/URLParser.cpp [hunk 6]
```diff
@@ -382,8 +564,22 @@ static Optional<DeprecatedString> parse_host(StringView input, bool is_not_speci
         }
     }
 
-    auto ipv4_host = parse_ipv4_address(ascii_domain);
-    return ipv4_host;
+    // 8. If asciiDomain ends in a number, then return the result of IPv4 parsing asciiDomain.
+    if (is_ascii_digit(ascii_domain[ascii_domain.length() - 1])) {
```
Comment (trflynn89) [inline AK/URLParser.cpp:568]: Is it possible for `input` to have been an empty string?

Comment (shannonbooth) [reply AK/URLParser.cpp:568]: No, step 3. of the spec has an assertion that input is not empty (which is later copied to asciiDomain

File: AK/URLParser.cpp (hunk 2 above)
Comment (shannonbooth) [reply AK/URLParser.cpp:177]: Thanks, good spot. Missed fixing it up in my copy-paste

File: AK/URLParser.cpp (hunk 1 above)
Comment (shannonbooth) [reply AK/URLParser.cpp:49]: thanks, appreciated - that does look better :^)

**Pull request merged**